File: src/main/java/com/netsensia/rivalchess/model/util/FenUtils.java
package com.netsensia.rivalchess.model.util;

import com.netsensia.rivalchess.model.Colour;
import com.netsensia.rivalchess.model.Square;
import com.netsensia.rivalchess.model.exception.IllegalFenException;
import com.netsensia.rivalchess.model.Board;
import com.netsensia.rivalchess.model.SquareOccupant;

public class FenUtils {

    private static final String FEN_START_POS = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";

    private FenUtils() {}

    public static String getStartPos() {
        return FEN_START_POS;
    }

    public static Board getBoardModel(final String fen) {

        final String[] fenParts = fen.split(" ");

        if (fenParts.length < 4) {
            throw new IllegalFenException(
                    "Expected at least 2 sections to FEN - board and mover"
            );
        }

        final Board.BoardBuilder boardBuilder = Board.builder();

        setBoardParts(boardBuilder, fenParts[0]);
        setMover(boardBuilder, fenParts[1]);
        setCastleFlags(boardBuilder, fenParts.length > 2 ? fenParts[2] : "kqKQ");
        setEnPassant(boardBuilder, fenParts.length > 3 ? fenParts[3] : "-");
        setHalfMoves(boardBuilder, fenParts.length > 4 ? fenParts[4] : "0");
        setFullMoves(boardBuilder, fenParts.length > 5 ? fenParts[5] : "0");

        return boardBuilder.build();

    }

    private static void setMover(final Board.BoardBuilder boardBuilder, final String boardPart) {
        if (boardPart.length() != 1) {
            throw new IllegalFenException("Unexpected error processing side to move");
        }

        final char mover = boardPart.toCharArray()[0];
        switch (mover) {
            case 'w': boardBuilder.withSideToMove(Colour.WHITE); break;
            case 'b': boardBuilder.withSideToMove(Colour.BLACK); break;
            default:
                throw new IllegalFenException("Invalid side to move: " + mover);
        }
    }

    private static void setEnPassant(final Board.BoardBuilder boardBuilder, final String boardPart) {
        if (boardPart.equals("-")) {
            boardBuilder.withEnPassantFile(-1);
            return;
        }
        if (boardPart.length() != 2) {
            throw new IllegalFenException("Unexpected error processing en passant part: " + boardPart);
        }

        final char enPassantFileAlgebraic = boardPart.toCharArray()[0];
        if (enPassantFileAlgebraic < 'a' || enPassantFileAlgebraic > 'h') {
            throw new IllegalFenException("Invalid en passant part: " + boardPart);
        }

        boardBuilder.withEnPassantFile(enPassantFileAlgebraic - 'a');
    }

    private static void setCastleFlags(final Board.BoardBuilder boardBuilder, final String boardPart) {
        boardBuilder.withIsWhiteQueenSideCastleAvailable(boardPart.contains("Q"));
        boardBuilder.withIsWhiteKingSideCastleAvailable(boardPart.contains("K"));
        boardBuilder.withIsBlackQueenSideCastleAvailable(boardPart.contains("q"));
        boardBuilder.withIsBlackKingSideCastleAvailable(boardPart.contains("k"));
    }

    private static void setHalfMoves(final Board.BoardBuilder boardBuilder, final String boardPart) {
        boardBuilder.withHalfMoveCount(Integer.parseInt(boardPart));
    }

    private static void setFullMoves(final Board.BoardBuilder boardBuilder, final String boardPart) {
        boardBuilder.withFullMoveCount(Integer.parseInt(boardPart));
    }

    private static void setBoardParts(final Board.BoardBuilder boardBuilder, final String boardPart) {

        final String[] rankParts = boardPart.split("/");

        if (rankParts.length != 8) {
            throw new IllegalFenException(
                    "Expected 8 ranks in board part of FEN - found " + rankParts.length
            );
        }

        for (int i=0; i<8; i++) {
            rankParts[i] = expand(rankParts[i]);
        }

        getBoardFromRank(boardBuilder, rankParts, 0);
    }

    private static String expand(String rankPart) {
        return rankPart
                .replace("1", "-")
                .replace("2", "--")
                .replace("3", "---")
                .replace("4", "----")
                .replace("5", "-----")
                .replace("6", "------")
                .replace("7", "-------")
                .replace("8", "--------");
    }

    private static void getBoardFromRank(
            final Board.BoardBuilder boardBuilder, final String[] rankParts, final int rank) {

        if (rank < 8) {
            collectRankParts(boardBuilder, rankParts, 0, rank);
            getBoardFromRank(boardBuilder, rankParts, rank+1);
        }

    }

    public static void collectRankParts(
            final Board.BoardBuilder boardBuilder, final String[] rankParts, final int file, final int rank) {

        if (file == 8) {
            return;
        }

        final char piece = rankParts[rank].toCharArray()[file];
        switch (piece) {
            case 'p' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BP); break;
            case 'P' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WP); break;
            case 'q' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BQ); break;
            case 'Q' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WQ); break;
            case 'r' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BR); break;
            case 'R' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WR); break;
            case 'b' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BB); break;
            case 'B' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WB); break;
            case 'n' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BN); break;
            case 'N' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WN); break;
            case 'k' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BK); break;
            case 'K' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WK); break;
            case '-' : boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.NONE); break;
            default:
                throw new IllegalFenException("Unexpected character in FEN: " + piece);
        }
        collectRankParts(boardBuilder, rankParts, file+1, rank);
    }

    public static String invertFen(String fen) {
        fen = fen.trim();

        fen = fen.replace(" b ", " . ");
        fen = fen.replace(" w ", " ; ");

        fen = fen.replace('Q', 'z');
        fen = fen.replace('K', 'x');
        fen = fen.replace('N', 'c');
        fen = fen.replace('B', 'v');
        fen = fen.replace('R', 'm');
        fen = fen.replace('P', ',');

        fen = fen.replace('q', 'Q');
        fen = fen.replace('k', 'K');
        fen = fen.replace('n', 'N');
        fen = fen.replace('b', 'B');
        fen = fen.replace('r', 'R');
        fen = fen.replace('p', 'P');

        fen = fen.replace('z', 'q');
        fen = fen.replace('x', 'k');
        fen = fen.replace('c', 'n');
        fen = fen.replace('v', 'b');
        fen = fen.replace('m', 'r');
        fen = fen.replace(',', 'p');

        fen = fen.replace(" . ", " w ");
        fen = fen.replace(" ; ", " b ");

        String[] fenParts = fen.split(" ");
        String[] boardParts = fenParts[0].split("/");

        String newFen =
                boardParts[7] + "/" +
                        boardParts[6] + "/" +
                        boardParts[5] + "/" +
                        boardParts[4] + "/" +
                        boardParts[3] + "/" +
                        boardParts[2] + "/" +
                        boardParts[1] + "/" +
                        boardParts[0];

        StringBuilder newFenBuilder = new StringBuilder(newFen);

        for (int i = 1; i < fenParts.length; i++) {
            if (i == 3) {
                newFenBuilder.append(" ").append(invertSquare(fenParts[i]));
            } else {
                newFenBuilder.append(" ").append(fenParts[i]);
            }
        }

        return newFenBuilder.toString();
    }

    private static String invertSquare(final String square) {
        if (square.equals("-")) {
            return square;
        }

        if (square.length() != 2) {
            throw new IllegalFenException("Invalid square reference " + square);
        }

        final char file = square.charAt(0);
        final char rank = square.charAt(1);

        final char newFile = (char)('h' - file + 'a');
        final char newRank = (char)('8' - rank + '1');

        return String.valueOf(newFile) + newRank;
    }
}

File: src/main/java/com/netsensia/rivalchess/model/util/FenUtils.kt
package com.netsensia.rivalchess.model.util

import com.netsensia.rivalchess.model.Board
import com.netsensia.rivalchess.model.Board.BoardBuilder
import com.netsensia.rivalchess.model.Colour
import com.netsensia.rivalchess.model.Square
import com.netsensia.rivalchess.model.SquareOccupant
import com.netsensia.rivalchess.model.exception.IllegalFenException
import com.netsensia.rivalchess.model.helper.SliderDirection
import java.util.stream.Collectors.toList

object FenUtils {

    @kotlin.jvm.JvmStatic
    fun getBoardModel(fen: String): Board {
        val fenParts = fen.trim().split(" ").toTypedArray()
        if (fenParts.size < 4) {
            throw IllegalFenException(
                    "Expected at least 2 sections to FEN - board and mover"
            )
        }
        val boardBuilder: BoardBuilder = Board.builder()
        setBoardParts(boardBuilder, fenParts[0])
        setMover(boardBuilder, fenParts[1])
        setCastleFlags(boardBuilder, if (fenParts.size > 2) fenParts[2] else "kqKQ")
        setEnPassant(boardBuilder, if (fenParts.size > 3) fenParts[3] else "-")
        setHalfMoves(boardBuilder, if (fenParts.size > 4) fenParts[4] else "0")
        setFullMoves(boardBuilder, if (fenParts.size > 5) fenParts[5] else "0")
        return boardBuilder.build()
    }

    private fun setMover(boardBuilder: BoardBuilder, boardPart: String) {
        if (boardPart.length != 1) {
            throw IllegalFenException("Unexpected error processing side to move")
        }
        when (val mover = boardPart.toCharArray()[0]) {
            'w' -> boardBuilder.withSideToMove(Colour.WHITE)
            'b' -> boardBuilder.withSideToMove(Colour.BLACK)
            else -> throw IllegalFenException("Invalid side to move: $mover")
        }
    }

    private fun setEnPassant(boardBuilder: BoardBuilder, boardPart: String) {
        if (boardPart == "-") {
            boardBuilder.withEnPassantFile(-1)
            return
        }
        if (boardPart.length != 2) {
            throw IllegalFenException("Unexpected error processing en passant part: $boardPart")
        }
        val enPassantFileAlgebraic = boardPart.toCharArray()[0]
        if (enPassantFileAlgebraic < 'a' || enPassantFileAlgebraic > 'h') {
            throw IllegalFenException("Invalid en passant part: $boardPart")
        }
        boardBuilder.withEnPassantFile(enPassantFileAlgebraic - 'a')
    }

    private fun setCastleFlags(boardBuilder: BoardBuilder, boardPart: String) {
        boardBuilder.withIsWhiteQueenSideCastleAvailable(boardPart.contains("Q"))
        boardBuilder.withIsWhiteKingSideCastleAvailable(boardPart.contains("K"))
        boardBuilder.withIsBlackQueenSideCastleAvailable(boardPart.contains("q"))
        boardBuilder.withIsBlackKingSideCastleAvailable(boardPart.contains("k"))
    }

    private fun setHalfMoves(boardBuilder: BoardBuilder, boardPart: String) {
        boardBuilder.withHalfMoveCount(boardPart.toInt())
    }

    private fun setFullMoves(boardBuilder: BoardBuilder, boardPart: String) {
        boardBuilder.withFullMoveCount(boardPart.toInt())
    }

    private fun setBoardParts(boardBuilder: BoardBuilder, boardPart: String) {
        val rankParts = boardPart.split("/").toTypedArray()
        if (rankParts.size != 8) {
            throw IllegalFenException(
                    "Expected 8 ranks in board part of FEN - found " + rankParts.size
            )
        }
        for (i in 0..7) {
            rankParts[i] = expand(rankParts[i])
        }
        getBoardFromRank(boardBuilder, rankParts, 0)
    }

    private fun expand(rankPart: String): String {
        return rankPart
                .replace("1", "-")
                .replace("2", "--")
                .replace("3", "---")
                .replace("4", "----")
                .replace("5", "-----")
                .replace("6", "------")
                .replace("7", "-------")
                .replace("8", "--------")
    }

    private fun getBoardFromRank(
            boardBuilder: BoardBuilder, rankParts: Array<String>, rank: Int) {
        if (rank < 8) {
            collectRankParts(boardBuilder, rankParts, 0, rank)
            getBoardFromRank(boardBuilder, rankParts, rank + 1)
        }
    }

    private fun collectRankParts(
            boardBuilder: BoardBuilder, rankParts: Array<String>, file: Int, rank: Int) {
        if (file == 8) {
            return
        }
        when (val piece = rankParts[rank].toCharArray()[file]) {
            'p' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BP)
            'P' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WP)
            'q' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BQ)
            'Q' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WQ)
            'r' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BR)
            'R' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WR)
            'b' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BB)
            'B' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WB)
            'n' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BN)
            'N' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WN)
            'k' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.BK)
            'K' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.WK)
            '-' -> boardBuilder.withSquareOccupant(Square.fromCoords(file, rank), SquareOccupant.NONE)
            else -> throw IllegalFenException("Unexpected character in FEN: $piece")
        }
        collectRankParts(boardBuilder, rankParts, file + 1, rank)
    }

    @kotlin.jvm.JvmStatic
    fun invertFen(fen: String): String {
        var invertedFen = fen
        invertedFen = invertedFen.trim { it <= ' ' }
        invertedFen = invertedFen.replace(" b ", " . ")
        invertedFen = invertedFen.replace(" w ", " ; ")
        invertedFen = invertedFen.replace('Q', 'z')
        invertedFen = invertedFen.replace('K', 'x')
        invertedFen = invertedFen.replace('N', 'c')
        invertedFen = invertedFen.replace('B', 'v')
        invertedFen = invertedFen.replace('R', 'm')
        invertedFen = invertedFen.replace('P', ',')
        invertedFen = invertedFen.replace('q', 'Q')
        invertedFen = invertedFen.replace('k', 'K')
        invertedFen = invertedFen.replace('n', 'N')
        invertedFen = invertedFen.replace('b', 'B')
        invertedFen = invertedFen.replace('r', 'R')
        invertedFen = invertedFen.replace('p', 'P')
        invertedFen = invertedFen.replace('z', 'q')
        invertedFen = invertedFen.replace('x', 'k')
        invertedFen = invertedFen.replace('c', 'n')
        invertedFen = invertedFen.replace('v', 'b')
        invertedFen = invertedFen.replace('m', 'r')
        invertedFen = invertedFen.replace(',', 'p')
        invertedFen = invertedFen.replace(" . ", " w ")
        invertedFen = invertedFen.replace(" ; ", " b ")
        val fenParts = invertedFen.split(" ").toTypedArray()
        val boardParts = fenParts[0].split("/").toTypedArray()
        val newFen = boardParts[7] + "/" +
                boardParts[6] + "/" +
                boardParts[5] + "/" +
                boardParts[4] + "/" +
                boardParts[3] + "/" +
                boardParts[2] + "/" +
                boardParts[1] + "/" +
                boardParts[0]
        val newFenBuilder = StringBuilder(newFen)
        for (i in 1 until fenParts.size) {
            if (i == 3) {
                newFenBuilder.append(" ").append(invertSquare(fenParts[i]))
            } else {
                newFenBuilder.append(" ").append(fenParts[i])
            }
        }
        return newFenBuilder.toString()
    }

    private fun invertSquare(square: String): String {
        if (square == "-") {
            return square
        }
        if (square.length != 2) {
            throw IllegalFenException("Invalid square reference $square")
        }
        val file = square[0]
        val rank = square[1]
        val newFile = ('h' - file + 'a'.toInt()).toChar()
        val newRank = ('8' - rank + '1'.toInt()).toChar()
        return newFile.toString() + newRank
    }

    private fun getEmptyCount(board: Board, square: Square): Int {

        if (board.getSquareOccupant(square) != SquareOccupant.NONE) {
            return 0;
        }

        return 1 + if (square.isValidDirection(SliderDirection.E))
            getEmptyCount(board, square.fromDirection(SliderDirection.E))
        else 0
    }

    private fun getRowString(board: Board, square: Square): String {

        val sb = StringBuilder()

        val squareOccupant = board.getSquareOccupant(square)

        if (squareOccupant != SquareOccupant.NONE) {
            sb.append(squareOccupant.toChar())
            if (square.isValidDirection(SliderDirection.E)) {
                sb.append(getRowString(board, square.fromDirection(SliderDirection.E)))
            }
        } else {
            val emptyToRight = if (square.isValidDirection(SliderDirection.E))
                getEmptyCount(board, square.fromDirection(SliderDirection.E))
            else 0
            sb.append(1 + emptyToRight)
            val newSquare = square.fromDirection(SliderDirection.E, emptyToRight)
            if (newSquare.isValidDirection(SliderDirection.E)) {
                sb.append(getRowString(board, newSquare.fromDirection(SliderDirection.E)))
            }
        }

        return sb.toString()
    }

    @kotlin.jvm.JvmStatic
    fun Board.getFen(): String {
        val sb = StringBuilder()
        for (square in listOf(Square.A8, Square.A7, Square.A6, Square.A5, Square.A4, Square.A3, Square.A2)) {
            sb.append(getRowString(this, square));
            sb.append("/")
        }
        sb.append(getRowString(this, Square.A1));
        sb.append(" ");
        sb.append(this.sideToMove.value)
        sb.append(" ")
        sb.append(if (this.isKingSideCastleAvailable(Colour.BLACK)) "k" else "")
        sb.append(if (this.isQueenSideCastleAvailable(Colour.BLACK)) "q" else "")
        sb.append(if (this.isKingSideCastleAvailable(Colour.WHITE)) "K" else "")
        sb.append(if (this.isQueenSideCastleAvailable(Colour.WHITE)) "Q" else "")
        sb.append(" ")
        sb.append((this.enPassantFile + 97).toChar())
        sb.append(if (this.sideToMove == Colour.WHITE) '6' else '3')
        sb.append(" ")
        sb.append(this.halfMoveCount)
        sb.append(" ")
        sb.append(this.fullMoveCount)

        return sb.toString()

    }
}